lib/storage/src/Upload.ts

```diff
@@ -0,0 +1,185 @@
+    this.uploadEvent = event;
+    super.on(event, listener);
+  }
+
+  async __doConcurrentUpload(dataFeeder: AsyncGenerator<RawDataPart>, id: number): Promise<void> {
+    for await (const dataPart of dataFeeder) {
+      if (this.uploadedParts.length > this.MAX_PARTS) {
+        throw new Error(
+          `Exceeded ${this.MAX_PARTS} as part of the upload to ${this.params.Key} and ${this.params.Bucket}.`
+        );
+      }
+
+      try {
+        let partResult = await this.client.send(
+          new UploadPartCommand({
+            ...this.params,
+            UploadId: this.uploadId,
+            Body: dataPart.data,
+            PartNumber: dataPart.partNumber,
+          })
+        );
+        if (this.abortSignal.signal.aborted) {
+          return;
+        }
+
+        this.uploadedParts.push({
+          PartNumber: dataPart.partNumber,
+          ETag: partResult.ETag,
+        });
+
+        this.bytesUploadedSoFar += byteLength(dataPart.data);
+        this.__notifyProgress({
+          loaded: this.bytesUploadedSoFar,
+          total: this.totalBytes,
+          part: dataPart.partNumber,
+          Key: this.params.Key,
+          Bucket: this.params.Bucket,
+        });
+      } catch (e) {
+        // on leavePartsOnError throw an error so users can deal with it themselves,
+        // otherwise swallow the error
+        if (this.leavePartsOnError) {
+          throw e;
+        }
+      }
+    }
+  }
+
+  async __doMultipartUpload(): Promise<ServiceOutputTypes> {
+    let createMultipartUploadResult = await this.client.send(new CreateMultipartUploadCommand(this.params));
+    this.uploadId = createMultipartUploadResult.UploadId;
+
+    // Set up data input chunks
+    const dataFeeder = getChunk(this.params.Body, this.partSize);
+
+    // Create and start concurrent uploads
+    for (let index = 0; index < this.queueSize; index++) {
+      const currentUpload = this.__doConcurrentUpload(dataFeeder, index);
+      this.concurrentUploaders.push(currentUpload);
+    }
+
+    // Create and start concurrent uploads
+    await Promise.all(this.concurrentUploaders);
+    if (this.abortSignal.signal.aborted) {
+      return {} as ServiceOutputTypes;
+    }
+
+    this.uploadedParts.sort((a, b) => a.PartNumber! - b.PartNumber!);
+    const completeMultipartUpload = await this.client.send(
+      new CompleteMultipartUploadCommand({
+        ...this.params,
+        Bucket: this.params.Bucket,
+        Key: this.params.Key,
+        UploadId: this.uploadId,
+        MultipartUpload: {
+          Parts: this.uploadedParts,
+        },
+      })
+    );
+    return completeMultipartUpload;
+  }
+
+  __notifyProgress(progress: Progress) {
+    if (this.uploadEvent) {
+      this.emit(this.uploadEvent, progress);
+    }
+  }
+
+  async __abortTimeout(abortSignal: AbortSignal): Promise<ServiceOutputTypes> {
+    return new Promise((resolve) => {
+      abortSignal.onabort = () => resolve({} as ServiceOutputTypes);
+    });
+  }
```

Comment:
abortTimeout should reject onabort.
```suggestion
  async __abortTimeout(abortSignal: AbortSignal): Promise<unknown> {
    return new Promise((reject) => {
      abortSignal.onabort = () => {
        const abortError = new Error("Upload aborted.");
        abortError.name = "AbortError";
        reject(abortError);
      };
    });
  }
```

Reply:
addressed